node_modules/oae-principals/lib/api.user.js

```diff
@@ -119,63 +126,275 @@ var createUser = module.exports.createUser = function(ctx, tenantAlias, displayN
+            //       user signs into the system.
+            log().debug({'tenant': tenant.alias, 'email': email, 'inviter': ctx.user().id}, 'Creating guest user on specific tenant');
+            return createUser(ctx, tenant.alias, displayName, opts, callback);
+        }
+
+        // We couldn't map an email address to an existing tenant. We'll have to create
+        // a user account on the guests tenant
+        var guestsTenant = TenantsAPI.getGuestsTenant();
+        log().debug({'email': email, 'inviter': ctx.user().id}, 'Creating guest user on guests tenant');
+        createUser(ctx, guestsTenant.alias, displayName, opts, function(err, user) {
+            if (err) {
+                return callback(err);
+            }
+
+            // Generate a token that will be sent to the user's email address
+            var token = _generateEmailToken();
+
+            // Store the token
+            PrincipalsDAO.storeEmailToken(token, user.id, function(err) {
+                if (err) {
+                    return callback(err);
+                }
+
+                // Send an email to the specified e-mail address
+                var data = {
+                    'tenant': guestsTenant,
+                    'invitee': user,
+                    'inviter': ctx.user(),
+                    'inviterDisplayName': ctx.user().displayName,
+                    'baseUrl': TenantsUtil.getBaseUrl(guestsTenant),
+                    'skinVariables': require('oae-ui').getTenantSkinVariables(guestsTenant.alias),
+                    'token': token
+                };
+                EmailAPI.sendEmail('oae-principals', 'invite', user, data, {'hash': token, 'ignoreVerification': true}, function(err) {
+                    if (err) {
+                        return callback(err);
+                    }
+
+                    return callback(null, user, true);
+                });
+            });
+        });
+    });
+};
+
+/**
+ * Given a set of strings, create (or retrieve) a guest account for the strings that
+ * are email addresses. If the string is not an email address, it's expected to be
+ * a user or group identifier. If this is not the case, a validation error will be
+ * returned.
+ *
+ * @param  {Context}        ctx                         Standard context object containing the current user and the current tenant
+ * @param  {String[]}       principals                  An array of principal ids to retrieve mixed with email address for which guest accounts should be created
+ * @param  {Function}       callback                    Standard callback function
+ * @param  {Object}         callback.err                An error that occurred, if any
+ * @param  {Object}         callback.principalIds       Object mapping each principal of `principals` to a `principalId`
+ */
+var getOrCreateGuestUsers = module.exports.getOrCreateGuestUsers = function(ctx, principals, callback) {
```

Comment:
Replacing this with a function that gets a set of principals and creates guest users along the way where necessary makes sense to me now.